### PR TITLE
Expose random_state as parameter for thresholders that use RNGs

### PR DESCRIPTION
**Pull request by CodeLionX**
I integrated PyThresh into our [TimeEval tool](https://github.com/HPI-Information-Systems/TimeEval/blob/main/timeeval/metrics/thresholding.py#L345) and I noticed that REGR builds a normal distribution without seeding the RNG. This leads to different results when calling it multiple times despite the same input:

```python
import numpy as np
from pythresh.thresholds.regr import REGR

scores = np.random.default_rng().random(1000)
regr = REGR(random_state=42)
test1 = regr.eval(scores)
test2 = regr.eval(scores)

# this raises an AssertionError
np.testing.assert_array_equal(test1, test2)
```

This PR lets REGR expose a `random_state` parameter to seed the NumPy RNG, similar to other thresholders.

---

**EDIT**

Common handling of `random_state` for all thresholders that use RNGs:

- all
- boot
- clust
- decomp
- dsn
- mcst
- ocsvm
- regr
- wind

All of them expose a `random_state` parameter to seed their RNGs (either used directly or through included libraries, such as sklearn) that is set by default to `1234`. This keeps the previous default behavior.

### CODE REVIEW

**Comment by KulikDM** [comment]
Hi @CodeLionX thanks for submitting this PR! 100%, the random state should have been set for this method. Currently, the rest of the other methods that have `random_state` set have the variable set statically within them (also to reduce the number of input variables). Therefore, it would probably be best to do so for the `REGR` thresholder. If you can change do this, I will push the PR through. Thanks again!

**Comment by CodeLionX** [comment]
hm 🤔 

I would actually argue the other way: All thresholders that use some kind of randomness should provide a way to change the seed, but provide a default one (reproducible default behavior). This does not increase the number of parameters the users have to set, but gives the API more flexibility. That would be similar to the NumPy, Pandas, or Scikit-learn APIs. E.g., it allows testing the influence of the randomness on the results and provides a sensible default behavior.
In this case, we should change the other thresholders to expose their `random_state` parameter as well. I identified the following thresholders:

- all
- boot
- clust
- decomp
- dsn
- mcst
- ocsvm
- wind

---

If you stick to your decision, we must also change the VAE-thresholder:

https://github.com/KulikDM/pythresh/blob/e89168aae8d9691abaf331d94f4b9d92782e2920/pythresh/thresholds/vae.py#L59-L60

Please let me know your opinion on it.

**Comment by KulikDM** [comment]
Hi @CodeLionX having the `random_state` statically set was done more so that the user wouldn't use it to manipulate the resulting contamination level (consistency-wise). However, you are right in that it would give the API more flexibility, and allow for the testing the influence of randomness on the results.  Also, yes the `VAE` thresholder would also require this parameter. It was more a personal choices I made when setting up this package. However, if you are willing to make the necessary changes to the thresholders that require this parameter, I am more than willing to merge this pull request. Thank you for your efforts so far!

**Comment by CodeLionX** [comment]
**Off-topic**: When working on `pythresh.thresholds.clust.CLUST`, I noticed the following fishy lines:

https://github.com/KulikDM/pythresh/blob/f846b805a6cabaeab46f5791d916d79bb9957ba6/pythresh/thresholds/clust.py#L242-L245

You might want to look into this.

**Comment by KulikDM** [comment]
@CodeLionX , thank you. Sorry I somehow missed your final message. This issue has now been fixed. Appreciate the find. Also to give you an update, exposing the `random_state` parameter was a good decision. In the PyThresh documentation under Benchmarks you will find the fruit of your efforts as I have added a section on the effects of randomness on thresholding (interesting results). Anyways, all the best!

**Comment by CodeLionX** [comment]
Happy new year! 🎉 

Nice analysis and definitely interesting results! Thanks for the update.

Maybe it would be a good idea to add a warning to the PyDoc of the seed-sensitive thresholders (`WIND`, `VAE`, `REGR`, `DSN`) with the hint to use it in an ensemble? I do not guess everyone will read the Benchmark-section of your documentation, but the class-doc is quite prominent in most tools.

**Comment by KulikDM** [comment]
Great idea, I'll add it as a hint in the Class docs as well.

Much appreciated!
